In this notebook we compare once again the 2 most promising algorithms from the models comparison and decide wich one implement and use for testing.
In particular we use the BagOfWord model from Preprocessing1.ipynb and we compare Logistic Regression and Ridge (with alpha = 20) that were the 2 algorithms that seemed to perform better.
In comparing them we will consider:
- Accuracy on the original distribution with hold-out
- Accuracy computed with 10 fold Cross Validation and the corresponding variance;
- Precision;
- Recall;
- F1Score;


In [26]:
string = "In this notebook we compare once again the 2 most promising algorithms from the models comparison and decide wich one implement and use for testing. \n In particular we use the BagOfWord model from Preprocessing1.ipynb and we compare Logistic Regression and Ridge (with alpha = 20) that were the 2 algorithms that seemed to perform better.\n In comparing them we will consider:\n - Accuracy on the original distribution with hold-out\n - Accuracy computed with 10 fold Cross Validation and the corresponding variance;\n - Precision;\n - Recall;\n - F1Score;"

In [29]:
print(string)

In this notebook we compare once again the 2 most promising algorithms from the models comparison and decide wich one implement and use for testing. 
 In particular we use the BagOfWord model from Preprocessing1.ipynb and we compare Logistic Regression and Ridge (with alpha = 20) that were the 2 algorithms that seemed to perform better.
 In comparing them we will consider:
 - Accuracy on the original distribution with hold-out
 - Accuracy computed with 10 fold Cross Validation and the corresponding variance;
 - Precision;
 - Recall;
 - F1Score;


In [1]:
print("comparison of Logistic Regression and Ridge (with alpha = 20) from Preprocessing1 BAgOfWord model")

comparison of Logistic Regression and Ridge (with alpha = 20) from Preprocessing1 BAgOfWord model


In [2]:
# dataframe management
import pandas as pd             

# numerical computation
import numpy as np



# seaborn can generate several warnings, we ignore them
import warnings 
warnings.filterwarnings("ignore")


from sklearn.model_selection import cross_val_score
import sklearn.metrics as mt
from sklearn import metrics
from sklearn.model_selection import train_test_split
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import recall_score


In [3]:
#Load the dataset from Preprocessinc1.ipynb
print("Loading the Dataset from Preprocessing1.ipynb.. ")
dataset = pd.read_csv("./BagOfWordDataSet1.csv")

Loading the Dataset from Preprocessing1.ipynb.. 


In [4]:
dataset.head()

Unnamed: 0  aaa  aadhaar  aadhar  aakash  aalto  aaron  aathar  abathroom  \
0           0    0        0       0       0      0      0       0          0   
1           1    0        0       0       0      0      0       0          0   
2           2    0        0       0       0      0      0       0          0   
3           3    0        0       0       0      0      0       0          0   
4           4    0        0       0       0      0      0       0          0   

   abbi   ...    zookeep  zoolog  zoologist  zoroastrian  zstsn  zubeen  \
0     0   ...          0       0          0            0      0       0   
1     0   ...          0       0          0            0      0       0   
2     0   ...          0       0          0            0      0       0   
3     0   ...          0       0          0            0      0       0   
4     0   ...          0       0          0            0      0       0   

   zuckerberg  zuni  zusak  TARGET  
0           0     0      0       0  
1           0     0      0       0  
2           0     0      0       0  
3           0     0      0       0  
4           0     0      0       0  

[5 rows x 10471 columns]

In [5]:
dataset = dataset.drop('Unnamed: 0', 1)
dataset.head()

aaa  aadhaar  aadhar  aakash  aalto  aaron  aathar  abathroom  abbi  \
0    0        0       0       0      0      0       0          0     0   
1    0        0       0       0      0      0       0          0     0   
2    0        0       0       0      0      0       0          0     0   
3    0        0       0       0      0      0       0          0     0   
4    0        0       0       0      0      0       0          0     0   

   abbrevi   ...    zookeep  zoolog  zoologist  zoroastrian  zstsn  zubeen  \
0        0   ...          0       0          0            0      0       0   
1        0   ...          0       0          0            0      0       0   
2        0   ...          0       0          0            0      0       0   
3        0   ...          0       0          0            0      0       0   
4        0   ...          0       0          0            0      0       0   

   zuckerberg  zuni  zusak  TARGET  
0           0     0      0       0  
1           0     0      0       0  
2           0     0      0       0  
3           0     0      0       0  
4           0     0      0       0  

[5 rows x 10470 columns]

In [6]:
y = dataset['TARGET']
variables = dataset.columns[dataset.columns!='TARGET']
y.head()

0    0
1    0
2    0
3    0
4    0
Name: TARGET, dtype: int64

In [7]:
X = dataset[variables]
X.head()

aaa  aadhaar  aadhar  aakash  aalto  aaron  aathar  abathroom  abbi  \
0    0        0       0       0      0      0       0          0     0   
1    0        0       0       0      0      0       0          0     0   
2    0        0       0       0      0      0       0          0     0   
3    0        0       0       0      0      0       0          0     0   
4    0        0       0       0      0      0       0          0     0   

   abbrevi  ...    zone  zookeep  zoolog  zoologist  zoroastrian  zstsn  \
0        0  ...       0        0       0          0            0      0   
1        0  ...       0        0       0          0            0      0   
2        0  ...       0        0       0          0            0      0   
3        0  ...       0        0       0          0            0      0   
4        0  ...       0        0       0          0            0      0   

   zubeen  zuckerberg  zuni  zusak  
0       0           0     0      0  
1       0           0     0      0  
2       0           0     0      0  
3       0           0     0      0  
4       0           0     0      0  

[5 rows x 10469 columns]

In [8]:
#splitting train and test set in a stratified manner
print("splitting train and test set in a stratified manner..")
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=123, stratify = y)

splitting train and test set in a stratified manner..


#### Now we can perform Oversampling on the training set:

In [9]:
from pandas import *
df_train= concat([X_train, y_train], axis=1, sort=False)
df_train.head()

aaa  aadhaar  aadhar  aakash  aalto  aaron  aathar  abathroom  abbi  \
8305    0        0       0       0      0      0       0          0     0   
3395    0        0       0       0      0      0       0          0     0   
1456    0        0       0       0      0      0       0          0     0   
1202    0        0       0       0      0      0       0          0     0   
3102    0        0       0       0      0      0       0          0     0   

      abbrevi   ...    zookeep  zoolog  zoologist  zoroastrian  zstsn  zubeen  \
8305        0   ...          0       0          0            0      0       0   
3395        0   ...          0       0          0            0      0       0   
1456        0   ...          0       0          0            0      0       0   
1202        0   ...          0       0          0            0      0       0   
3102        0   ...          0       0          0            0      0       0   

      zuckerberg  zuni  zusak  TARGET  
8305           0     0      0       0  
3395           0     0      0       0  
1456           0     0      0       0  
1202           0     0      0       0  
3102           0     0      0       0  

[5 rows x 10470 columns]

In [10]:
# Class count
count_class_0, count_class_1 = df_train.TARGET.value_counts()
print("negative class instances:")
print(count_class_0)
print("positive class instances:")
print(count_class_1)

print("percentange of positive: " + str(round(count_class_1*100/(count_class_0 + count_class_1),2)))

negative class instances:
8441
positive class instances:
559
percentange of positive: 6.21


In [11]:
# Divide by class
df_class_0 = df_train[df_train['TARGET'] == 0]
df_class_1 = df_train[df_train['TARGET'] == 1]

In [12]:
print("performing Oversampling..")
df_class_1_over = df_class_1.sample(count_class_0, replace=True)
train_over = pd.concat([df_class_0, df_class_1_over], axis=0)

count_class_0, count_class_1 = train_over.TARGET.value_counts()
print("negative class instances:")
print(count_class_0)
print("positive class instances:")
print(count_class_1)

print("percentange of positive: " + str(round(count_class_1*100/(count_class_0 + count_class_1),2)))

performing Oversampling..
negative class instances:
8441
positive class instances:
8441
percentange of positive: 50.0


In [13]:
#split again in X_train, y_train
y_train = train_over['TARGET']
variables = train_over.columns[dataset.columns!='TARGET']
X_train = train_over[variables]

In [14]:
def PrintConfusionMatrix(model, true_y, predicted_y, positive=1, negative=-1):
    cm = confusion_matrix(true_y,predicted_y)
    print("\t"+str(model.classes_[0])+"\t"+str(model.classes_[1]))
    print(str(model.classes_[0]) + "\t",cm[0][0],"\t",cm[0][1])
    print(str(model.classes_[1]) + "\t",cm[1][0],"\t",cm[1][1])

In [15]:
def F1Score ( precision, recall):
    f1 = 2 * precision * recall / (precision + recall)
    return f1

In [16]:
from sklearn.linear_model import LogisticRegression

In [17]:
print()
print("Logistic Regression")
ln = LogisticRegression(solver = 'liblinear')
ln.fit(X_train, y_train);
predictions_logistic = ln.predict(X_train)
# check  training accuracy
import sklearn.metrics as mt
acc = mt.accuracy_score(predictions_logistic,y_train)
print( "Training Accuracy: " + str(acc))


Logistic Regression
Training Accuracy: 0.9919440824546855


In [18]:
from sklearn.model_selection import cross_val_score
predictions_logistic = ln.predict(X_test)
accTest = mt.accuracy_score(predictions_logistic,y_test)
print("Test Accuracy on original distribution: " + str(accTest))

scores = cross_val_score(ln, X_train, y_train, cv=10)
#print(scores)
print("performing Cross Validation...")
print("Test Accuracy with CV: " + str(scores.mean()))
print("var: " + str(scores.var()))

Test Accuracy on original distribution: 0.931
performing Cross Validation...
Test Accuracy with CV: 0.9748240275947166
var: 3.030649478427743e-05


In [19]:
print("Confusion Matrix - Logistic Regression")
PrintConfusionMatrix(ln, y_test, predictions_logistic)

Confusion Matrix - Logistic Regression
	0	1
0	 895 	 43
1	 26 	 36


In [20]:
print("Precision %3.2f" % precision_score(y_test,predictions_logistic))
print("Recall    %3.2f" % recall_score(y_test,predictions_logistic))
print("F1    %3.2f" % F1Score( precision_score(y_test,predictions_logistic), recall_score(y_test,predictions_logistic)))

Precision 0.46
Recall    0.58
F1    0.51


In [21]:
from sklearn.linear_model import RidgeClassifier

In [22]:
print()
print("Ridge:")

    
#Train
clf = RidgeClassifier(alpha=20)
clf.fit(X_train, y_train)
print("alpha: " + str(20))
print()
predictions_ridge = clf.predict(X_train) 
acc = mt.accuracy_score(predictions_ridge,y_train)
print( "Training Accuracy: " + str(acc))
    
#Test on original distribution
predictions_ridge = clf.predict(X_test)
accTest = mt.accuracy_score(predictions_ridge,y_test)
print("Test Accuracy on original distribution: " + str(accTest))

#CV
scores = cross_val_score(clf, X_train, y_train, cv=10)
print("performing Cross Validation...")
print("Test Accuracy with CV: " + str(scores.mean()))
print("var: " + str(scores.var()))


    
print("Precision on test %3.2f" % precision_score(y_test,predictions_ridge))
print("Recall on test  %3.2f" % recall_score(y_test,predictions_ridge))
f1score = F1Score( precision_score(y_test,predictions_ridge), recall_score(y_test,predictions_ridge))
print("F1    %3.2f" % f1score )

print()


Ridge:
alpha: 20

Training Accuracy: 0.9809856652055444
Test Accuracy on original distribution: 0.931
performing Cross Validation...
Test Accuracy with CV: 0.9690186909335651
var: 2.932901630574358e-05
Precision on test 0.46
Recall on test  0.60
F1    0.52



### Comparison: 
(results can slightly vary for different runs)

#### Accuracy on original Distribution
Logistic Regression: 0.926 <br>
Ridge: 0.93

##### Accuracy with CV
Both Accuracy are pretty high but keep in mind that the original Datased was skewed and before performing Oversampling on the training set we had only 6.20 % of positive instances. Again this means that a classifier that predicts everithing as negative will achieve an Accuracy of 93.8%.
Here the Logistic Regression classifier shows an Accuracy of  0.975% evaluated with 10 Folds CrossValidation.
On the other hand Ridge has an Accuracy of  0.968% evaluated with the same criteria.
Both variances are small and not particularly significative.
Finally remember that these accuracies are computed on the oversampled distribution so they are not so significative.

##### Precision, Recall and F1Score
Here we notice that Ridge has both Precision and Recall greater than the ones of Logistic Regression (and consequently also the f1Score).
Logistic Regression: Precision  0.44; Recall    0.58; F1    0.50.
Ridge:               Precision  0.46; Recall    0.61; F1    0.53.
Again, Precision focuses on the percentage of correctly classified positive examples and Recall focuses on the percentage of positively classified examples with respect to the number of existing ones.

$precision = \frac{TP}{TP+FP}$

$recall = \frac{TP}{TP+FN}$

$ F1 = \frac{2 * Precison * Recall}{Precison + Recall}$

In this contex, where the goal is to correctly identify positive instances (Insincere Questions) these 2 metrics are more significative than a mere Accuracy and this is the reason why will pick the Ridge classifier as our model.

#### At this point let's save our model in order to test its perforfamance on the test somewhere else without training it everytime.

In [1]:
print("final model implementation: Ridge classisier trained on the oversampled dataset from Preprocessing1.py")

final model implementation: Ridge classisier trained on the oversampled dataset from Preprocessing1.py


#### save the model:

In [24]:
from joblib import dump
print("Saving the final model..")
dump(clf, 'FinalModel.joblib')
print("done!")

Saving the final model..
done!


##### save also the test set to be load in the final notebook where we compare the final model with the initial baselines:

In [30]:
### we also save the test set
df_test= concat([X_test, y_test], axis=1, sort=False)
df_test.head()

aaa  aadhaar  aadhar  aakash  aalto  aaron  aathar  abathroom  abbi  \
4130    0        0       0       0      0      0       0          0     0   
740     0        0       0       0      0      0       0          0     0   
7389    0        0       0       0      0      0       0          0     0   
6133    0        0       0       0      0      0       0          0     0   
1873    0        0       0       0      0      0       0          0     0   

      abbrevi   ...    zookeep  zoolog  zoologist  zoroastrian  zstsn  zubeen  \
4130        0   ...          0       0          0            0      0       0   
740         0   ...          0       0          0            0      0       0   
7389        0   ...          0       0          0            0      0       0   
6133        0   ...          0       0          0            0      0       0   
1873        0   ...          0       0          0            0      0       0   

      zuckerberg  zuni  zusak  TARGET  
4130           0     0      0       0  
740            0     0      0       0  
7389           0     0      0       0  
6133           0     0      0       0  
1873           0     0      0       1  

[5 rows x 10470 columns]

In [31]:
df_test.to_csv(r'./testSet.csv')